library/x509_crt.c

```diff
@@ -1435,32 +1571,179 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
 }
 #endif /* MBEDTLS_FS_IO */
 
+/*
+ * OtherName ::= SEQUENCE {
+ *      type-id    OBJECT IDENTIFIER,
+ *      value      [0] EXPLICIT ANY DEFINED BY type-id }
+ *
+ * HardwareModuleName ::= SEQUENCE {
+ *                           hwType OBJECT IDENTIFIER,
+ *                           hwSerialNum OCTET STRING }
+ *
+ * NOTE: we currently only parse and use otherName of type HwModuleName,
+ * as defined in RFC 4108.
+ */
+int mbedtls_x509_get_other_name( const mbedtls_x509_sequence *subject_alt_name,
+                                 mbedtls_x509_san_othername *other_name )
+{
+    int ret;
+    size_t len;
+    unsigned char *p = subject_alt_name->buf.p;
+    const unsigned char *end = p + subject_alt_name->buf.len;
+    mbedtls_x509_buf cur_oid = { 0, 0, NULL };
+
+    if( subject_alt_name->buf.tag != ( MBEDTLS_ASN1_CONTEXT_SPECIFIC | 0 ) )
+    {
+        /*
+         * The given subject alternative name is not of type "othername".
+         */
+        return( MBEDTLS_ERR_X509_BAD_INPUT_DATA );
+    }
+
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
+                                      MBEDTLS_ASN1_OID ) ) != 0 )
+        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS + ret );
+
+    cur_oid.tag = MBEDTLS_ASN1_OID;
+    cur_oid.p = p;
+    cur_oid.len = len;
+
+    /*
+     * Only HwModuleName is currently supported.
+     */
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_ON_HW_MODULE_NAME, &cur_oid ) != 0 )
+    {
+        return( MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE );
+    }
+    p += len;
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
```

Comment:
Perhaps we should parse the whole extension earlier (in x509_get_subject_alt_name())? i mean, is there a situation where a certificate is parsed, then verification of the chain (somehow) succeeds and later when we try to load the other name we fail to parse it?

Reply:
I did this at first, but I saw it adds some complexity and too much API change.
In addition, this is more according to #2478 

>i mean, is there a situation where a certificate is parsed, then verification of the chain (somehow) succeeds and later when we try to load the other name we fail to parse it?

Well, yes, if the `otherName` is not of type `hardwareModuleName`, but I don't see a problem with that.

Reply:
ok

Reply:
perhaps the naming should be different. But this is actually semi parsing. We *do* parse the certificate into different `mbedtls_x509_sequence` types, we just don't *interpret* them